nostr-client-plus/src/bin/publisher.rs

```diff
@@ -102,37 +118,31 @@ async fn run() -> Result<()> {
         }
     });
 
-    let timestamp_now = chrono::Utc::now().timestamp() as u64;
-
-    let entries = left_anti_join(&collection, "finished_jobs", limit_publish).await?;
+    // Let's figure out number of classifications and pow to send out
+    let classification_count = (limit_publish as f64 * classification_job_percentage) as i64;
+    println!("Plan to publish {} classification jobs", classification_count);
 
-    // Get type job id and name
-    let job_type = JobType::Classification.job_type();
-    let job_type_str = JobType::Classification.as_ref();
+    // Now let's fetch all classification entries needed
+    // Note there is a side effect where the entries here can have less entries than expected in case
+    // of all entries are fetched
+    let mut entries: Vec<Document> = Vec::new();
+    // Only fetch classification jobs when the number of jobs needed is larger than 0
+    if classification_count > 0 {
+        entries = left_anti_join(&collection, "finished_jobs", classification_count).await?;
+    }
+    println!("Actually fetched {} classfications to publish", entries.len());
 
+    // Next, send out events
     let mut jobs_sent = 0;
-    for entry in entries {
-        let entry: RawDataEntry = from_document(entry)?;
-        println!("Entry {}", entry._id.to_string());
-        let header = PayloadHeader {
-            job_type,
-            raw_data_id: entry._id,
-            time: timestamp_now,
-        };
-        let payload = NewJobPayload {
-            header,
-            kv_key: entry.r2_key,
-            config: None,
-            validator: "default".to_string(),
-            classifier: "default".to_string(),
-        };
-        let event = UnsignedEvent::new(
-            client.sender(),
-            timestamp_now,
-            Kind::NEW_JOB,
-            vec![vec!["t".to_string(), job_type_str.to_string()]],
-            serde_json::to_string(&payload).expect("Payload serialization failed"),
-        );
+    for i in 0..limit_publish {
+        // Since all classifications to send out is inside entries
+        // if i is larger than len of that, it must be PoW job
+        let mut entry: Option<RawDataEntry> = None;
+        if i < entries.len() as i64 {
+            // Save to unwrap, item at i should always exists since it's less than the vec length
+            entry = from_document(entries.get(i as usize).unwrap().clone())?;
```

Comment:
OK, I do understand why you cloned here. It's not the end of the world because publisher is not such a time-critical component, but there was another way of doing this without cloning, because you are copying elements that we will never use again and throw away. They could have been consumed.

In a running service, cloning here would have not been acceptable.